### DISCUSSION TITLE
Add/sync remote folder to knowledge base

### DISCUSSION THREAD

**Discussion by mahenning** [open-webui/open-webui#7933] [Q&A]
There was a previous `Scan` option, where you can sync with a directory to get the content into `open-webui`. Previously, it was a fixed path and I could use a volume to mount documents. [It got removed](https://github.com/open-webui/open-webui/discussions/6013) in favour of a more fine-grained knowledge construct.

My question is, how do I sync with a remote folder again? I run `open-webui` in a docker container on a headless VM, and access the UI remotely on another machine. When I select "Sync directory" in a given knowledge base, I can only select a folder on my local computer. Thats not where the documents are, they are on the server VM.
- is this option intended to sync a folder live, so that when a new file is loaded into the folder, it is automatically added to the knowledge base? Or do I have to manually re-sync the folder if stuff is added?
- how do I set a path to a folder on the remote VM (where the documents are) for the knowledge base?

I have found [the API docs](https://docs.openwebui.com/getting-started/advanced-topics/api-endpoints#-retrieval-augmented-generation-rag) for document upload endpoint, but as far as I understood it is a one-time (not live-sync) uploading and then (tediously) adding them to a specific knowledge base. I doesn't look like a replacement for the old `Scan` option.

**Comment by muntedcrocodile** [comment]
Was looking for a feature like this it's disappointing to discover its been removed. If you try and sync the same folder again it will discard all embedding and start again instead of only updating the ones that have changed. The scan option sounds like exactly what i want.

**Reply by bryongaskin** [reply]
Yeah, the scan feature was nice.  If you change anything in terms of the vector database / directory it wipes everything out.  The files are still sitting uploads folder in the underling linux structure but Web Openai doesn't see  them.  Without gong to a command line and writing code, it seems to me the best option is that you copy uploads folder to a folder on your windows PC, create knowledge and do a directory synch. The reason a say this is if you wrote any context text in Open Webui, that doesn't get included in the resync.  It seems like there should be subfolder under Uploads for each knowledge.  I think it would be easier to reconstruct / manage

**Comment by gputrain** [comment]
I've been digging through various discussions and would love to better understand the community's current approaches to knowledge management, especially around keeping folders in sync.

The previous Scan feature allowed mounting volumes and syncing directories, supposedly kept knowledge in sync?

Key questions for the community:

- How are you currently handling incremental updates to your knowledge bases?
- What approaches have you found effective for maintaining sync between file systems and vector stores?
- For those using external tools, what patterns have emerged as particularly robust?

The rationale behind removing the Scan feature isn't immediately clear to me, but I imagine there were valid technical considerations. Understanding these would help us address any underlying concerns while still meeting the use case.

One particularly compelling aspect of incremental updates is resource efficiency - processing only changed files rather than rebuilding the entire vector store. This becomes crucial as knowledge bases grow larger. (Obsidian vault for example)

**Reply by noah-schroeder** [reply]
I am trying to figure out the same thing. I have a large knowledge base (i.e,. hours to sync) and I don't want to clear it every time I want to add a few new documents. Is there a way anyone has found to automatically scan the folder and just add/remove content instead of wiping/rebuilding from 0?

**Reply by hades200082** [reply]
Right now we don't really have much choice but to manually re-sync things via the UI.

The ideal solution for us would be as follows...

We currently use WikiJS to manage our knowledgebases. Each client has their own instance which our team keep up to date with knowledge about that client's solutions. We also have a separate instance for all of our internal knowledge, and another for the company handbook, etc.

Each WikiJS instance is configured to sync it's content with a Github repo.

Ideally we would be able to create a new "Knowledge" in Open WebUI and point it at a git repo. Changes in the repo should be automatically pulled and updated in the vector store.

**Reply by mahenning** [reply]
You can write a script that does the following:
- scan your upload directory for current files (also with timestamps if you want)
- call the API (with user key) <your-open-webui>/api/v1/knowledge/<knowledge-id>
You get a json which (among other things) includes all files with names and "created at"/"updated at" timestamps for these files.
Then you can look up if:
- A new file is in the folder -> upload it to the knowledge
- A file in the folder has a newer timestamp than the one in the knowledge -> update it (delete + upload)
- (A file is in the knowledge that does not exist in the upload folder anymore -> delete it)

All list / upload / delete are API calls. Put everything together in a script, and if you want let it run as a cronjob.

**Reply by hades200082** [reply]
Even in Docker? 

It would be a great feature to have included in OI though. No key needed... just connect OI to github API using a PAT and let it take care of it.

**Reply by hades200082** [reply]
Some possible use-cases...

**1 - Internal teams like ours**
The team's day-to-day updates of knowledgebases/wiki's already creates git commits in the relevant repos. Linking those repos to OI would just tie in what we already maintain with the AI knowledge.

**2 - Official docs**
Many tools and services use github to maintain their official docs in markdown format. 

- [Microsoft .NET](https://github.com/dotnet/docs/tree/main/docs)
- [NextJS](https://github.com/vercel/next.js/tree/v15.5.3/docs)
- [even Open WebUI itself](https://github.com/open-webui/docs/tree/main/docs)
- etc..

If we could point at these folders in the git repos as well it would make a very powerful knowledge base AI for developers, systems architects, etc.

**Comment by panicsteve** [comment]
I arrived here looking to see if there was a way that I could have, say, a folder of Markdown files that Open WebUI would periodically scan for modifications and update a knowledge base accordingly (while also being smart about only re-vectorizing files that have changed, etc.)

It sounds like the removed "Scan" feature did what I want, but possibly not in an automated way. Any plans to bring this feature back in any form?

I found this: https://github.com/open-webui/cookbook/blob/main/knowledge/add-to-knowledge.ipynb but it would be great if this were just an integrated feature. I think?

**Comment by YutaroYamada** [comment]
To improve the performance of the sync directory, we are considering the following design:

* When adding files to the knowledge base, raw data will be stored in a corresponding dedicated folder.
* When the sync directory is executed, it will be compared with the previous sync directory run, and only the differences will be added to or deleted from the knowledge base.
* For differential comparison, we will use watchdog's snapshot feature. Each time the sync directory completes, this snapshot will be saved. During the next run, the previous snapshot will be compared with the current snapshot to detect additions, deletions, and modifications, and appropriate processing will be performed for each.
* Currently, records holding files in the knowledge base do not have a unique column for easy file identification. We will add the file path as a new unique column. This will enable the new implementation of a file deletion API.

With this configuration, the performance of the sync directory can be improved. Additionally, it will facilitate the future implementation of features such as automatic synchronization by periodically monitoring the directory.


@tjbck 
Do you have any concerns or feedback regarding the implementation of these? I would greatly appreciate your comments.

**Reply by tyagivasu** [reply]
Hi @YutaroYamada, is there any feature issue already open for this that tracks it?

**Reply by YutaroYamada** [reply]
@tyagivasu 
No, I didn't create an Issue because I checked the project rules and it says to post new features to "discussion" before submitting a PR.

As for whether I've already started development, I haven't had a lot of dedicated time recently, so it's unlikely I'll be able to make progress this month. I'd be more than happy for you to go ahead and implement or submit a PR yourself.

**Comment by hades200082** [comment]
The ideal solution for us would be as follows...

We currently use WikiJS to manage our knowledge bases. Each client has their own instance which our team keep up to date with knowledge about that client's solutions. We also have a separate instance for all of our internal knowledge, and another for the company handbook, etc.

Each WikiJS instance is configured to sync it's content with a Github repo.

Ideally, we would be able to create a new "Knowledge" in Open WebUI and point it at a git repo. Changes in the repo should be automatically pulled and updated in the vector store.

**Reply by meganoob1337** [reply]
https://github.com/castai/openwebui-content-sync
I think this should fit your needs. it can sync a repo, local files and confluence with the OWUI Knowledge base. you can specify a sync interval as it keeps track of if a page in confluence has changed, probably the same for github, havent tried that part yet

**Reply by hades200082** [reply]
That looks promising. 

Since Open WebUI runs as a docker compose it would be nice if that could be added to the compose instead of needing a different orchestration architecture.

**Reply by meganoob1337** [reply]
There is a Docker image, you would need to look at what you need to mount to get it to run though. 

in general you just need to mount the config file , data folder and add the env variables for the api keys for the adapters and OWUI 
env variables are documented and config aswell. data dir is just /data i think. 
you could add an Issue aswell and i might add some documentation ( i made 2 PRs to the Project , but am not owner/maintainer)

**Reply by meganoob1337** [reply]
Docker images are available now, I think also a little documentation, my jira adapter also got merged

**Reply by flefevre** [reply]
Great project. Is it compatible with other sources such as gitlab?

**Reply by flefevre** [reply]
Have you seen the news in Google which covers some part of the usecase

https://developers.googleblog.com/introducing-the-developer-knowledge-api-and-mcp-server/

**Reply by SchneiderSam** [reply]
@hades200082 are you happy with wikijs? I don't like that only one person is maintaining it and the maintainer is very busy 😬

Actually we are using an obsidian vault for internal wiki but I don't like the rights management there. Not so clear who changed what...

**Comment by f18m** [comment]
I wonder if there was any progress on this topic...

**Comment by duys-cc** [comment]
Bumping this thread, so it doesn't close out. Any progress on this?

**Reply by aprudnev** [reply]
We need it too. It's actually way better if we can have directory mounted next to OWUI and configure it to keep it in sync, vs using GUI from the personal system (which is thousands miles away). API... API is terrible. You can not just call API and ask to resync if file changed, and you can not even call API to add file to KB - you must first call API to scan fine, then you wait get file id and add file id into KB...

And of course many serious usages will require OWUI to index LOCAL folder not folder available thru the web server on GUI side. Aside of other things, it can not be automated, and it is slow (as file travel many times between source , browser and server, instead of been mounted on the server as SMB or NFS share).

**Comment by Classic298** [comment]
Once my PR gets merged or the feature gets added, proper sync will be introduced and these new api endpoints could be used to sync any folder from anywhere.

**Reply by SchneiderSam** [reply]
@Classic298 woow wow wow. Thanks so much for working on this. 🔥

**Reply by SchneiderSam** [reply]
It would be great if existing knowledge bases could continue to be connected to an external source, so that when you have a large amount of knowledge, you can still deploy it in a targeted way — meaning you can control which model accesses which specific knowledge.

In other words, having an interface that allows you to pull from or crawl external knowledge bases would be huge. At the moment, I handle this manually by uploading all the data myself. But when data changes within the team, it becomes very time-consuming to keep track of everything. If this process were automated, that would be a game changer.

What we are currently trying to avoid, however, is putting all knowledge into a single, centralized knowledge base.

**Reply by SchneiderSam** [reply]
@Classic298 can you share the PR to see it?

**Reply by Classic298** [reply]
Look at my prs

**Reply by d-shehu** [reply]
I'm a bit confused by the response and the whole "sync directory" concept. I looked through the comments from the PRs but not the code from Classic298.

Is the goal to support auto-syncing with some kind of file watch so OpenWebUI's vector DB is incrementally updated as docs are added/removed/updated?

I think users would be happy to organize their documents into collections of directories and then create a KB per collection. I started to code something similar on my end but I haven't gotten around to figuring out how to integrate here. I was also thinking of a tags based approach where collections could be created from sets of tags for greater flexibility. Ideally tags could be derived from scraping / auto-indexing of the docs.

Sorry, I'm not trying to jump into solution space but I feel the KB in OpenWebUI needs some TLC. And I don't understand the long term evolution?

For reference, I was hoping for an open-source solution that's similar to Zettlab's Semantic search. Thanks.

**Reply by slimhk45** [reply]
Can't wait for it. Knowledge base is useless if we can't add local folders by syncing.
